### PR TITLE
Expose AppContext.TargetFrameworkName in the contracts

### LINKED ISSUE #3049: Add a switch to AppContext that identifies the current framework
Scenario
===

We would like to enable loading references from nuget packages in Roslyn scripts. A sample script could look like follows:

```
#r "System.IO.FileSystem.4.0.0.nuget"

var stream = new FileStream(@"C:\temp\file");
```

The app running this script would use Roslyn Hosting APIs to compile and execute the script code at runtime. Our script engine would 
1) download the nuget package to local cache, if it's not already available 
2) find the implementation assembly(ies) in the package [1]
3) compile the script against these assemblies [2]
4) execute the script (which loads the assemblies)

The question we need to answer in step 2) is what is the identifier of the current framework, since the nuget may contain implementation assemblies targeting multiple platforms/frameworks. 

Proposal
===

We propose that the CLR host exposes the framework identifier to the application thru AppContext.

We will assume that the name of the framework is "net46" if we don't find the value in AppContext. Roslyn Scripting API requires 4.6+ FX on desktop, so if the value is properly set on future releases of Desktop FX and by all CoreCLR hosts we will always find the right one.

Notes
===
[1] If the package doesn't have impl assemblies then we fall back to GAC lookup.
[2] The script engine always compiles against runtime/implementation assemblies (very much like Reflection.Emit). Contract assemblies are usually not available on non-developer machines. We don't require all dependencies to be loaded thru nuget packages, which would give us access to contract assemblies.

### PR DESCRIPTION
**Pull request by AlexGhiondea**
The implementation was already checked in, this is exposing the API in the contracts and sets up packaging for it.

Fixes #3049

### CODE REVIEW

**Comment by davidfowl** [comment]
Finally :smile:

File: src/System.AppContext/ref/System.AppContext.cs [hunk 1]
```diff
@@ -12,5 +12,6 @@ public static partial class AppContext
         public static string BaseDirectory { get { return default(string); } }
         public static void SetSwitch(string switchName, bool isEnabled) { }
         public static bool TryGetSwitch(string switchName, out bool isEnabled) { isEnabled = default(bool); return default(bool); }
+		public static string TargetFrameworkName { get { return default(string); } }
```
Comment (davidfowl) [inline src/System.AppContext/ref/System.AppContext.cs]: nit: spacing

Comment (AlexGhiondea) [inline src/System.AppContext/ref/System.AppContext.cs]: Fixed!

**Comment by stephentoub** [comment]
Tests?

**Comment by AlexGhiondea** [comment]
@stephentoub the TFN as implemented returns a static string. Testing the TFN is not something that is trivial to add as we need to have both an actual implementation that reads the value from the assembly (like we do in Desktop) and a way to generate assemblies with different TFNs.

There are 2 issues on my plate now to handle tests.
1. Port the existing AppContext to GitHub (#851)
2. Figure out the way to test AppContext in a portable way (across all the supported platforms) (#5383)

Unless you strongly object I am planning to merge this is as-is (once the builds pass) and follow-up with the other issues.

**Comment by stephentoub** [comment]
Ok. Thanks.

**Comment by AlexGhiondea** [comment]
Something strange is happening with the build... that, or the log is broken. 

I am seeing things like this:
```
dotnet_corefx\centos7.1_debug_bld_prtest\packages/Microsoft.argetingPack.NetFramework.v4.6
```
Note the missing 'T' from Microsoft.TargetingPack

or 
```
d:\j\wokspace\dotnet_corefx\centos7.1_debug_bld_prtest\packages/Microsoft.TargetingPack.NetFramework.v4.6
```
Note the missing 'k' in d:\j\workspace

@ericstj @mmitche any idea what might be going on?

**Comment by ericstj** [comment]
1.  Windows_NT Debug Build and Test - seems to have failed due to an intermittently failing network test @CIPop 
2. Innerloop Windows_NT Release Build and Test - seems to have failed due to skipped test threshold (28) I see a few tests skipping because they weren't running elevated.  Is that expected @mmitche?
3. Ubuntu.* & Centos.* failed to build the package, could be a problem with the pkg or src .builds file.  /cc @weshaggard 
```
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(460,5): error : File d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\bin\AnyOS.AnyCPU.Debug\System.AppContext\net462\System.AppContext.dll does not exist, ensure you have built libraries before building the package. [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(460,5): error : File d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\bin\AnyOS.AnyCPU.Debug\System.AppContext\net462\System.AppContext.dll does not exist, ensure you have built libraries before building the package. [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error : Error when creating nuget package from d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\bin\pkg\specs\System.AppContext.nuspec. System.IO.FileNotFoundException: File not found: 'd:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\bin\AnyOS.AnyCPU.Debug\System.AppContext\net462\System.AppContext.dll'. [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error :    at NuGet.PackageBuilder.AddFiles(String basePath, String source, String destination, String exclude) [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error :    at NuGet.PackageBuilder.PopulateFiles(String basePath, IEnumerable`1 files) [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error :    at NuGet.PackageBuilder.ReadManifest(Stream stream, String basePath, IPropertyProvider propertyProvider) [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error :    at NuGet.PackageBuilder..ctor(String path, String basePath, IPropertyProvider propertyProvider, Boolean includeEmptyDirectories) [d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\src\System.AppContext\pkg\System.AppContext.pkgproj]
11:52:12 d:\j\workspace\dotnet_corefx\ubuntu_debug_bld_prtest\Tools\Packaging.targets(897,5): error :    at Microsoft.DotNet.Build.Tasks.Packaging.NuGetPack.Execute() 
```

**Comment by stephentoub** [comment]
FYI, I've already submitted a PR to disable that networking test.  Most of the members of that test class are all disabled now, likely with the same or a similar root cause, and there are several open issues for it.

**Comment by CIPop** [comment]
@stephentoub @ericstj  The failure signature doesn't match the intermittent test failures we've been seeing in the past:

```
MESSAGE:
Assert.True() Failure\r\nExpected: True\r\nActual: False
+++++++++++++++++++
STACK TRACE:
at System.Net.Security.Tests.NegotiateStreamStreamToStreamTest.NegotiateStream_StreamToStream_Authentication_EmptyCredentials_Fails() in d:\j\workspace\dotnet_corefx\windows_nt_debug_prtest\src\System.Net.Security\tests\FunctionalTests\NegotiateStreamStreamToStreamTest.cs:line 155
```

This is:
                Assert.Equal(TokenImpersonationLevel.Identification, client.ImpersonationLevel);

@AlexGhiondea Is the test passing if you build locally?

**Comment by stephentoub** [comment]
> The failure signature doesn't match the intermittent test failures we've been seeing in the past

This one has actually failed a bunch over the last day, in various PRs and main builds.  But you're right that there isn't an issue dedicated to this signature... I assumed that since you'd closed one or more of the other issues for the other NegotiateStream tests that I should just map this to the remaining one as well, rather than open another issue.  Apologies if that was a bad assumption.

**Comment by CIPop** [comment]
@stephentoub Thanks for explaining. Opened #5397 to track this.

@AlexGhiondea feel free to disable the test part of this PR referencing the issue above.

**Comment by mmitche** [comment]
@CIPop Yes we don't run elevated on purpose (sudo isn't the default, it's bad practice, etc.)  However we do run outerloop in elevated mode.

**Comment by CIPop** [comment]
@mmitche Did something in the CI cause this breaking change? In the interest of keeping this PR comments focused to the changed code, please reply in #5397 instead.

**Comment by AlexGhiondea** [comment]
I have fixed the issue that was causing the build break -- the targeting pack needed to be explicitly set to 4.6 for the net462 group. 

If the networking test continues to fail I will go ahead and disable it.

File: src/System.AppContext/src/System.AppContext.csproj [hunk 2]
```diff
@@ -5,9 +5,11 @@
     <AssemblyName>System.AppContext</AssemblyName>
     <ProjectGuid>{5522BAFC-E2FF-4896-993A-401DDEDFD85F}</ProjectGuid>
     <ClsCompliant>true</ClsCompliant>
-    <AssemblyVersion>4.0.1.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)'!='netcore50'">true</IsPartialFacadeAssembly>
-    <PackageTargetFramework Condition="'$(PackageTargetFramework)' == ''">dotnet5.4</PackageTargetFramework>
+    <PackageTargetFramework Condition="'$(PackageTargetFramework)' == ''">dotnet5.6</PackageTargetFramework>
+    <!-- The following line needs to be removed once we have a targeting pack for 4.6.2 -->
+    <TargetingPackNugetPackageId Condition="'$(TargetGroup)'=='net462'">Microsoft.TargetingPack.NETFramework.v4.6</TargetingPackNugetPackageId>
```
Comment (ericstj) [inline src/System.AppContext/src/System.AppContext.csproj]: I would have used the 4.6.1 tp instead.  But I guess it doesn't matter in your case.

**Pull request merged**